TypeScript strict mode support + docs

dac09 opened the pull request:
### Background
If you turn on TypeScript strict mode, (modify tsconfig.json)

```diff
{
  "compilerOptions": {
    "noEmit": true,
    "allowJs": true,
+    "strict": true,
```
Redwood projects currently have a bunch of errors, with no clear path to a solution. 

This PR is a combination of documentation, exporting and adding more types so that users opting-in to strict mode can have an error free project. Note that we are _not_ enabling strict mode by default.

### Changes in this PR

**Generate stricter Resolver definitions in strict mode**
- [x] When strict mode is turned on, make "args" of each of the resolvers (i.e. service functions) non-optional
- [x] Document changes to service tests to remove errors

**`src/lib/auth.ts`**
- [x] Documentation on how to get rid of the errors related to roles checks in src/lib/auth 
- [x] ~Add types for getCurrentUser for non-dbAuth~ - this really is too hairy

**dbAuth**

- [x] Make `DbAuthOptions` and `DbAuthSession` available for users to import in their project. 
- [x] Updates default generated templates in src/lib/auth and src/functions/auth to use these types
- [x] Allow generics, so the "user" in dbAuth can be typed
- [x] Documentation on how to use the generics
- ~Add a line to setup auth dbAuth output on "using dbAuth with TS"~ generator output is already too verbose

**Relation resolvers**
When you have a model with a relation, and you generate the sdl or service with the `--crud` flag, relational resolvers are generated. 

- [x] Changes the generator templates to use optional chaining to access `gqlArgs`. 
**Reasoning:** 
a) Although these are resolvers, they can be used outside the graphql context - where gqlArgs aren't passed.
b) The alternative is to ask users to modify all their service tests. Its ugly and confusing and unnecessary. 
- [x] Update [graphql resolver docs](https://redwoodjs.com/docs/graphql#redwoods-resolver-args)

**Prisma vs Graphql null types**
- [x] Document that Prisma differentiates between null and undefined (https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined) and recommend Orta's suggestion to use `dnull` in [this section](https://redwoodjs.com/docs/services)
- [x] Add example using dnull
- [x] Create issue for adding dnull to Redwood templates by default


**Avoid optional resolvers types**
This is a codegen config change (see graphqlCodegen.ts)
- [x] Make generated Resolver types non-optional, so they do not complain in strict mode


**Improve return type of Scenarios**
- [x] Expose `ScenarioData` type that lets you define exactly what the output of a scenario is
- [x] Add backlink [here](https://redwoodjs.com/docs/testing#scenarios) to scenario typing docs

**Service tests with chained calls**
- [x] Change template for service/scenario tests to tell TS to pipe down

**Prisma `findUnique` return types**
❌ No solution, just have to document



Closes #5481
Closes #5463
Closes #5261

dac09 commented:
Related:  #5463

and https://github.com/redwoodjs/redwood/issues/5261

dthyresson commented on packages/cli/src/commands/generate/scaffold/templates/components/Name.tsx.template:
```diff
@@ -23,15 +28,15 @@ const formatEnum = (values: string | string[] | null | undefined) => {
 }
 
-const timeTag = (datetime) => {
+const timeTag = (datetime?: string) => {
```
So, out of Prisma and the query, the date time value is a string and not a proper date?

Right, the ISO JSON value. This type here definitely helps to recognize that it's a string and not a date.

dthyresson commented on packages/cli/src/commands/generate/scaffold/templates/components/Name.tsx.template:
```diff
@@ -83,15 +94,15 @@ const ${singularPascalName} = ({ ${singularCamelName} }) => {
       <nav className="rw-button-group">
         <Link
-          to={routes.${editRouteName}({ id: ${singularCamelName}.id })}
+          to={routes.${editRouteName}({ id: ${singularCamelName}?.id })}
```
If this is a text template, it cares that `${singularCamelName}?.id` the template replacement text might be null? Interesting.

dac09 replied:
This isn't to do with the template actually. it just adds the ? after the name.

For example it would be `Post?.id`

dthyresson commented on packages/cli/src/commands/generate/scaffold/templates/components/NameCell.tsx.template:
```diff
@@ -16,7 +16,7 @@ export const Loading = () => <div>Loading...</div>
 
 export const Failure = ({ error }: CellFailureProps) => (
-  <div className="rw-cell-error">{error.message}</div>
+  <div className="rw-cell-error">{error?.message}</div>
```
Have to say, and maybe this is "strict" but given that this component would be shown only if there is an error, seems odd that needs the `?` since isn't its presence why the component is being rendered.

dac09 replied:
Yeah its a bit weird - but Apollo says these errors are optional, even when it errors out. That's probably because there's some sort of case where it doesn't return an error case

dthyresson commented on packages/cli/src/commands/generate/service/templates/service.ts.template:
```diff
@@ -30,6 +30,6 @@ export const delete${singularPascalName}: MutationResolvers['delete${singularPas
-export const ${singularPascalName}: ${singularPascalName}Resolvers  = {<% relations.forEach(relation => { %>
-  ${relation}: (_obj, { root }) => db.${singularCamelName}.findUnique({ where: { id: root.id } }).${relation}(),<% }) %>
+export const ${singularPascalName}: Partial<${singularPascalName}Resolvers>  = {<% relations.forEach(relation => { %>
+  ${relation}: (_obj, gqlArgs) => db.${singularCamelName}.findUnique({ where: { id: gqlArgs?.root?.id } }).${relation}(),<% }) %>
```
So, now we can access other arguments besides root.

See: https://graphql.org/learn/execution/#root-fields-resolvers

> args The arguments provided to the field in the GraphQL query.

dac09 replied:
We could always do this, it was just desctructured before. Any thoughts on the name? Best I could come up with  was`gqlArgs`

dthyresson reviewed (commented):
@dac09 I took a first pass and left some comments -- more questions, observations rather than actionable. The templates makes sense - there were just a few places where the `?` was used than surprised me, but perhaps that's what strict demands.

dac09 commented on packages/cli/src/commands/generate/cell/__tests__/__snapshots__/cell.test.js.snap:
```diff
@@ -14,7 +14,7 @@ export const Loading = () => <div>Loading...</div>
 
 export const Failure = ({ error }) => (
-  <div style={{ color: 'red' }}>Error: {error.message}</div>
+  <div style={{ color: 'red' }}>Error: {error?.message}</div>
```
True true

dac09 replied:
I'm not going to change this one actually DT - the only reason the error is optional is because of the Apollo typing - I don't even know in what case error might be null. In 99% of cases, errror will be there... so I think no need to change it.

dac09 commented:
Assigning you guys for a review :) 

- @jtoar - particularly on docs please, I know you often like phrasing things differently, feel free to change things!
- @Tobbe - for the typescript changes, and the strict mode docs if they make sense?
- @cannikin - could I get a review from you by trying out the generators and stuff, if it all works ok for you in the JS world?

jtoar commented on docs/docs/typescript/generated-types.md:
```diff
@@ -0,0 +1,157 @@
+
+If you're curious, you can find the generated types in the `.redwood/types`, `web/types/graphql.d.ts`, and `api/types/graphql.d.ts` directories. Broadly speaking, Redwood generates the following types:
+
+1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, Cells, and layouts on the web side
```
Is there more than what's enumerated here? Maybe the "mirror" types include the api side too? For a simple project with a Post model and dbAuth setup, this is what I'm seeing:

```
.redwood/types/mirror/
├── web/
│  └── src/
│     └── pages/
│        ├── NotFoundPage/
│        │  └── index.d.ts
│        └── FatalErrorPage/
│           └── index.d.ts
└── api/
   └── src/
      ├── services/
      │  └── posts/
      │     └── index.d.ts
      └── directives/
         ├── skipAuth/
         │  └── index.d.ts
         └── requireAuth/
            └── index.d.ts
```

dac09 replied:
```suggestion
1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, Cells, and layouts on the web side - and services, libs, etc. on the api side
```

jtoar commented on docs/docs/typescript/generated-types.md:
```diff
@@ -0,0 +1,157 @@
+1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, Cells, and layouts on the web side
+2. types based on your queries and mutations on the web side (in `web/types/graphql.d.ts`)
+3. types for resolvers based on your SDLs on the api side (in `api/types/graphql.d.ts`)
+4. types for testing, `currentUser`, etc.
```
I've seen all three used and I'm a little confused if they're different: is it `currentUser`, `CurrentUser`, or CurrentUser / current user (i.e. not a code thing at all?)

jtoar replied:
I.e. since `CurrentUser` is capitalized, is it a type? Whereas `currentUser` is a variable somewhere? And then current user is just the concept?

dac09 replied:
They're interchangable really. But `currentUser` refers to the object/property usually, `CurrentUser` refers to the type.

But it gets tricky, when you say "you can find this in your CurrentUser" - where it means both

jtoar commented on docs/docs/typescript/generated-types.md:
````diff
@@ -0,0 +1,157 @@
+}
+```
+
+These types help you by making sure you're returning an object in the shape of what you've defined in your SDL. Note that these types expect you to return the _complete_ type that you've defined in your SDL. You can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.
````
I don't quite understand the last two sentences in this paragraph.

dac09 replied:
What I meant to say is - "normally typing these things in graphql is a royal pain in the butt, because you can't just return a DateTime from Prisma (because graphql wants a string for a Date field).... but with Redwood, its all good, and these QueryResolver types already do this mapping for you" 

But I agree doesn't read too well.

jtoar commented on docs/docs/typescript/generated-types.md:
```diff
@@ -0,0 +1,157 @@
+
+:::tip Using VSCode?
+
+As a part of type generation, the [VSCode GraphQL extension](https://marketplace.visualstudio.com/items?itemName=GraphQL.vscode-graphql) configures itself based on the merged schema Redwood generates in `.redwood/schema.graphql`.
+You can configure it further in `graphql.config.js` at the root of your project.
```
This is only related cause the extensions consumes the generated types right?

dac09 replied:
Ummm the extension consumes the schema, but if you configure the graphql.config.js file, you can also tell the extension to include certain headers (e.g. for auth) and make queries right in vscode

jtoar commented on docs/docs/typescript/strict-mode.md:
```diff
@@ -0,0 +1,239 @@
+### Roles checks for CurrentUser in `src/lib/auth`
+
+When you setup auth, Redwood includes some template code for handling roles with the `hasRole` function.
+While Redwood does runtime checks to make sure it doesn't access roles if it doesn't exist, TypeScript in strict mode will highlight errors, depending on
```
depending on...? 👀

jtoar commented on docs/docs/typescript/strict-mode.md:
```diff
@@ -0,0 +1,239 @@
+
+## Enabling strict mode
+
+Enable strict mode by setting `strict` to true in `web/tsconfig.json` and `api/tsconfig.json`:
```
Should we mention `scripts/tsconfig.json` too?

dac09 replied:
Sure

jtoar commented on docs/docs/typescript/strict-mode.md:
```diff
@@ -0,0 +1,239 @@
+
+:::info Why this happens in strict mode
+
+In strict mode, Redwood generates different types for `QueryResolvers` and `MutationResolvers`, where the first argument—`args`—is no longer optional.
```
This used to be "args is no longer non-optional". That was too many mental cycles for me, but did I get it right as "no longer optional"? Or is it required?

dac09 replied:
Whoops 😆, yeah this is correct. Args is required.

jtoar commented on docs/docs/typescript/utility-types.md:
```diff
@@ -0,0 +1,222 @@
+| :----------------- | :-------------------------------------------------------------------------------------------------- |
+| `PrismaCreateType` | (Optional) the type imported from Prisma's create operation that goes into the "data" key           |
+| `TName`            | (Optional) the name or names of the models in your scenario                                         |
+| `TKey`             | (Optional) the keys in your scenario. These are really only useful while you write out the scenario |
```
There's some confusion between `TKey` vs `TKeys` for `defineScenario` and `ScenarioData` and I wasn't sure which was supposed to which.

jtoar replied:
It looks like `ScenarioData` takes `TKeys`, but the name of the generics in the src for `defineScenario` are a bit differently-named from what was written up here:

```ts
    PrismaCreateType extends { data: any },
    ModelName extends string | number | symbol = string | number | symbol,
    KeyName extends string | number | symbol = string | number | symbol
```

dac09 replied:
Updated so that the generic names are the same!

jtoar commented on docs/docs/typescript/strict-mode.md:
```diff
@@ -0,0 +1,239 @@
+Depending on your auth provider—i.e., anything but dbAuth—because it could change based on your account settings (if you include roles or other metadata), we can't know the shape of your decoded token at setup time.
+So you'll have to make sure that the `getCurrentUser` function is typed.
+
+To help you get started, the comments above the `getCurrentUser` function describe its parameters' types. We recommend typing `decoded` without using imported types from Redwood, as this may be a little too generic!
```
I'm a little confused here cause it seems like we say not to use imported types from Redwood, but one of the examples uses an imported type from Redwood?

dac09 replied:
Yeah I included both, to show how you can achieve it both ways. While importing the type "works" - its not terribly useful in actual use, which is why I'm suggesting to just type it yourself

jtoar commented on docs/docs/typescript/generated-types.md:
````diff
@@ -0,0 +1,157 @@
+}
+```
+
+The types for both `useAuth().currentUser` on the web side and `context.currentUser` on the api side will be the same—the `MyCurrentUser` interface.
````
I updated the copy—is this still correct?

dac09 replied:
Ummm.. I think its important that we point out that the type _comes_ from the `getCurrentUser` function

dac09 replied:
Actually, I covered it above. Change looks good

jtoar commented on docs/docs/typescript/generated-types.md:
````diff
@@ -0,0 +1,157 @@
+}
+```
+
+These types will also be handled automatically. But if you're returning a different Prisma model, you may need to write your own resolver type, as the type generator won't know how to map the Prisma type to the GraphQL return type.
````
Just to clarify, what you're saying is you have to do this when the union is two Prisma models right?

dac09 replied:
Yes that's right.

dac09 replied:
So if you return something like `union CandyResult  = Candy | Chocolate` - because the typegenerator doesn't handle this usecase (I think)

jtoar reviewed (changes requested):
Really awesome work Danny! 🚀

I took the liberty of committing to this branch—hope that was ok! Happy to revert anything. I've got a few more questions though. There were some things I wasn't sure about and needed clarification on. I commented on specific lines where I could.

Another thought I had while reviewing: I feel like we should try combining this all into one doc. We only split up auth and deployment because they were monstrously long. This doesn't feel so long as to warrant four separate docs yet. I'm happy to do the work of merging everything together, but what do you think?

dac09 commented on docs/docs/typescript/generated-types.md:
```diff
@@ -0,0 +1,146 @@
+
+config:
+  namingConvention:
+    typeNames: change-case-all#upperCase
```
Yeah this is the correct syntax. No idea about how yaml handles commets, maybe its just when its with a space?

dac09 commented on packages/internal/src/generate/graphqlCodeGen.ts:
```diff
@@ -124,7 +125,7 @@ function getPluginConfig() {
     prismaModels = localPrisma.ModelName
     Object.keys(prismaModels).forEach((key) => {
-      prismaModels[key] = `.prisma/client#${key} as Prisma${key}`
+      prismaModels[key] = `@prisma/client#${key} as Prisma${key}`
```
Both work, because prisma client is actually aliased to ./prisma in node_modules I think (that's where the generated types go) 

I just thought this was better, because it respects the yarn/node aliasing rather than overlook it.

Tobbe commented on __fixtures__/test-project/web/src/components/Contact/ContactForm/ContactForm.tsx:
```diff
@@ -7,10 +7,21 @@ import {
+}
+
+const ContactForm = (props: ContactFormProps) => {
+  const onSubmit = (data: Record<string, any>) => {
```
I think we should have much better types here. And to that end I wrote this PR: https://github.com/redwoodjs/redwood/pull/5867

dac09 replied:
Good stuff! I don’t know forms that well, so didn’t dig in.

Tobbe reviewed (commented):
I have some bigger-picture things I need to discuss with the team before I can finish reviewing this. I've asked @jtoar for some time on the meeting on Tuesday.

dac09 commented on __fixtures__/test-project/api/src/services/contacts/contacts.test.ts:
```diff
@@ -37,7 +39,7 @@ describe('contacts', () => {
 
   scenario('updates a contact', async (scenario: StandardScenario) => {
-    const original = await contact({ id: scenario.contact.one.id })
+    const original = (await contact({ id: scenario.contact.one.id })) as Contact
```
You’re probably right - My focus wasn’t so much on assessing what the generators generate, more on satisfying TS. The type casting here is fine though, because the test will fail if contact is null. This is what discussed on the call a few weeks ago - but happy to try a different way, but am not aware of the original reasoning.

dac09 replied:
I just realised, because Id Is auto generated by prisma, so we don’t know the Id ahead of time.

Tobbe replied:
> The type casting here is fine though, because the test will fail if contact is null. This is what discussed on the call a few weeks ago
Thanks for jogging my memory, I remember that now 👍

dac09 commented on packages/eslint-config/shared.js:
```diff
@@ -122,6 +122,7 @@ module.exports = {
         '@typescript-eslint/camelcase': 'off',
         'no-use-before-define': 'off',
         '@typescript-eslint/no-use-before-define': 'off',
+        '@typescript-eslint/prefer-namespace-keyword': 'off',
```
Yes - we just had some weird conflicting rules here.

cannikin commented on packages/cli/src/commands/generate/scaffold/templates/components/Names.tsx.template:
```diff
@@ -27,19 +30,20 @@ const formatEnum = (values: string | string[] | null | undefined) => {
-  let output = text
-  if (text && text.length > MAX_STRING_LENGTH) {
+const truncate = (value: string | number) => {
+  let output = value.toString()
```
This is blowing up if the DB field is `NULL` since `toString()` doesn't exist on `null`.

Maybe something like this instead?

```suggestion
  let output = value ? value.toString() : ''
```

dac09 replied:
Does this work for you @cannikin ? `  let output = value?.toString() ?? ''`

cannikin commented:
Not sure if this is expected, but after creating a test project that's converted to JS, there's still an `api/types/graphql.d.ts` and `web/types/graphql.d.ts` files left lying around? Could be confusing to have those laying around if your project isn't TS?

<img width="302" alt="image" src="https://user-images.githubusercontent.com/300/177425738-fc3418bb-7f0f-4116-83ed-bc2535d7d720.png">

cannikin commented on packages/cli/src/commands/generate/scaffold/templates/components/EditNameCell.tsx.template:
```diff
@@ -1,4 +1,9 @@
+} from 'types/graphql'
 
 import { navigate, routes } from '@redwoodjs/router'
 import type { CellSuccessProps, CellFailureProps } from '@redwoodjs/web'
```
This line gets removed when converting to JS, but then the import above get a red squiggle because of the empty line: 

<img width="653" alt="image" src="https://user-images.githubusercontent.com/300/177426570-615c6145-3879-46b0-b361-7cb33081129e.png">

dac09 commented:
> Not sure if this is expected, but after creating a test project that's converted to JS, there's still an api/types/graphql.d.ts and web/types/graphql.d.ts files left lying around. Could be confusing to have those laying around if your project isn't TS?

Fair point - I haven't changed the behaviour here though. I can look at disabling generating these types in another PR (don't want to expand the scope of this PR).

The reason we still generate _some_ types - is that under the hood they still help out with JS projects, with autocompletion, etc. in certain cases.

dac09 commented:
@Tobbe based on the issue you brought up the other day I think I've identified a shortcoming in the resolverFn, we _do_ need to keep it as a `Partial` unfortunately, because the mapped prisma types don't take into account which of the fields in the SDL are optional.

But... I do have a suggestion for your particular case. If you are modifying a mapped type i.e. An SDL model that also exists in Prisma, but you're addiing additional fields to your SDL model - you can just define a different type, for example:

```graphql

  type Post {
    id: Int!
    title: String!
    body: String!

  }

  type PostWithAdditionalFields {
    id: Int!
    title: String!
    body: String!
    bazinga: Boolean # <-- new field
  }


  type Mutation {
    updatePost(id: Int!, input: UpdatePostInput!): PostWithAdditionalFields! @requireAuth

```

This will still generate the `MutationResolvers['updatePost']` with your new type, but ofcourse if you use DateTimes, that's not going to be so easy to use.

cannikin commented on __fixtures__/test-project/web/src/components/Contact/Contacts/Contacts.tsx:
```diff
@@ -27,19 +30,20 @@ const formatEnum = (values: string | string[] | null | undefined) => {
-  let output = text
-  if (text && text.length > MAX_STRING_LENGTH) {
+const truncate = (value: string | number) => {
+  let output = value.toString()
```
Don't forget the same fix should be present in the text-project! :)

dac09 replied:
Thanks, regen-ing test project

cannikin commented on __fixtures__/test-project/web/src/components/Post/Posts/Posts.tsx:
```diff
@@ -27,19 +30,20 @@ const formatEnum = (values: string | string[] | null | undefined) => {
-  let output = text
-  if (text && text.length > MAX_STRING_LENGTH) {
+const truncate = (value: string | number) => {
+  let output = value.toString()
```
And here!

Tobbe commented on docs/docs/typescript/strict-mode.md:
```diff
@@ -0,0 +1,239 @@
+- but For Prisma, `null` is a value, and `undefined` means "do nothing"
+
+This is covered in detail in [Prisma's docs](https://www.prisma.io/docs/concepts/components/prisma-client/null-and-undefined), which we strongly recommend reading.
+But the gist of it is that, for Prisma's create and update operations, you have to make sure `null`s are converted to `undefined` from your GraphQL mutation inputs.
```
The way I read this it sounds as if this would be needed even if we're not in strict mode. Should we call this out somewhere in our "regular" docs?

Tobbe replied:
Do you know of a good place to put this, so that JS and non-strict-mode TS devs can find this info too? Maybe somewhere in our general "Service" docs? Or in our GraphQL docs?

dac09 replied:
Nope - I think we need to explore our options more on this. Perhaps making dnull part of the template, but really is out of scope for this PR.

Tobbe replied:
> Nope - I think we need to explore our options more on this. Perhaps making dnull part of the template, but really is out of scope for this PR.

Sure. Should I write up an issue, or do you have a TODO in your own notes for this somewhere already?

Tobbe replied:
Created an issue https://github.com/redwoodjs/redwood/issues/6016

Tobbe commented on packages/cli/src/commands/generate/scaffold/templates/components/Names.tsx.template:
```diff
@@ -27,19 +30,20 @@ const formatEnum = (values: string | string[] | null | undefined) => {
-  if (text && text.length > MAX_STRING_LENGTH) {
+const truncate = (value: string | number) => {
+  let output = value?.toString() ?? ''
+  if (value && output?.length > MAX_STRING_LENGTH) {
```
Here's where the change I mentioned way up top would go, if you want to implement it

dac09 replied:
I've refactored it, should make a little more sense now (hopefully)

Tobbe reviewed (commented):
Wow Danny! Great work. Together with the docs this is awesome.

Tobbe commented on packages/api/src/functions/dbAuth/DbAuthHandler.ts:
```diff
@@ -149,20 +160,20 @@ type Params = {
   password?: string
   method: AuthMethodNames
-  [key: string]: unknown
+  [key: string]: any
```
The answer is: Because further down, in `_createUser` we destructure `this.params` and store most of it in `...userAttributes`. We then pass `userAttributes` to `signup.handler`, and that function doesn't handle `unknown`.

So now I have a proposal to get rid of this now
```suggestion
```

And instead use `UserAttributes`.
```
type Params = UserAttributes & {
  username?: string
  password?: string
  method: AuthMethodNames
}
```
(notice how it makes a union of `UserAttributes` and the other attributes listed).

This is what I'm most unsure about, because now we've said that all extra param keys can only have string values.  This should be fine, because we already assign the extra keys to `Record<string, string>` in the `_createUser` function. But please let me know if you don't agree

dac09 replied:
I think you were meant to push up some changes here? Mind resolving this conversation? I don't quite follow what you were saying here...

Tobbe replied:
I'll do a followup PR instead. Makes it easier to reason about my proposed changes when we can see them in isolation

Tobbe commented on packages/api/src/functions/dbAuth/DbAuthHandler.ts:
```diff
@@ -775,6 +787,9 @@ export class DbAuthHandler {
 
       return newUser
     }
+
```
Wanted to comment on a line a few lines up, but GH wouldn't let me, so I post it here instead

```diff
- const newUser = await this.options.signup.handler({
+ const newUser = await this.options.signup.handler<TUser>({
```

And again, like 50 lines below this, I wanted to suggest this change
```diff
- _loginResponse(user: Record<string, any>, statusCode = 200) {
+ _loginResponse(user: Record<string, unknown>, statusCode = 200) {
```

Tobbe reviewed (commented):
Wanted to see if I could get rid of some `any`s in this file. I *think* I did it, but please an extra look. Also, I haven't tried to actually use this new version of DbAuthHandler

dac09 commented on packages/api/src/functions/dbAuth/DbAuthHandler.ts:
```diff
@@ -52,7 +52,7 @@ interface DbAuthHandlerOptions {
    */
   forgotPassword: {
-    handler: (user: Record<string, unknown>) => Promise<any>
+    handler: (user: TUser) => any
```
Have you tried these unknowns in an actual project @Tobbe ? It will degrade the experience to a halt I think - with errors like "obj isn't assignable to type unknown" etc.

This isn't an unknown, its a genuine any. if you look at the description Rob's provided in the auth.ts template.

Tobbe replied:
I have made a bunch of these changes locally on your branch already. If you want I can push them, and we can work from there. Or you can do them from scratch on your own 🙂 
Or you don't agree at all, and then we can discuss first 🙂

dac09 replied:
This is a great suggestion! Feel free to push!

Tobbe replied:
Skipping this for now. I got side-tracked into types for auth on the web side. I/we can come back to this in another PR

dac09 commented on packages/api/src/functions/dbAuth/DbAuthHandler.ts:
```diff
@@ -109,7 +109,7 @@ interface DbAuthHandlerOptions {
      * from `useAuth()`
      */
-    handler: (signupHandlerOptions: SignupHandlerOptions) => Promise<any>
+    handler: (signupHandlerOptions: SignupHandlerArgs) => any
```
Again, this is a genuine any.

dac09 replied:
>  Whatever you want to happen to your data on new user signup. Redwood will
 check for duplicate usernames before calling this handler. At a minimum
 you need to save the `username`, `hashedPassword` and `salt` to your
 user table. `userAttributes` contains any additional object members that
 were included in the object given to the `signUp()` function you got
 from `useAuth()`.

 If you want the user to be immediately logged in, return the user that
 was created.

 If this handler throws an error, it will be returned by the `signUp()`
 function in the form of: `{ error: 'Error message' }`.

 If this returns anything else, it will be returned by the
 `signUp()` function in the form of: `{ message: 'String here' }`

dac09 commented on packages/api/src/functions/dbAuth/DbAuthHandler.ts:
```diff
@@ -149,20 +160,20 @@ type Params = {
 }
 
-export class DbAuthHandler {
+export class DbAuthHandler<TUser extends Record<string | number, any>> {
```
This will severely degrade the experience, as soon as you try to access any property of the user inside the handlers.

Using unknowns here forces people to either specify the generic. There's no reason for us to be so prescriptive here. If they want additional typesafety, they can pass the generic. If they don't want that level of typesafety, they leave it as is.

It's important that we let the user choose the level of complexity they want to introduce in their code with TS, and they have the option to - rather than fight the framework.

dac09 commented on __fixtures__/test-project/web/src/components/Contact/Contacts/Contacts.tsx:
```diff
@@ -27,19 +30,20 @@ const formatEnum = (values: string | string[] | null | undefined) => {
-  if (text && text.length > MAX_STRING_LENGTH) {
+const truncate = (value: string | number) => {
+  let output = value?.toString() ?? ''
+  if (value && output?.length > MAX_STRING_LENGTH) {
```
I think the logic is correct, but there's probably a better way to write this. 

We only want to go into the `if` block if `value` exists. If `value` doesn't exist, `output` would still get an empty string, and we'd be substring-ing for no reason.

dac09 commented on packages/api/src/functions/dbAuth/DbAuthHandler.ts:
```diff
@@ -15,7 +15,7 @@ import { normalizeRequest } from '../../transforms'
 import { decryptSession, extractCookie, getSession } from './shared'
 
-interface DbAuthHandlerOptions {
+export interface DbAuthHandlerOptions<TUser = Record<string | number, any>> {
```
I don't see any benefit here from using unknown. `any` allows the user to make the decision whether they want to pass the user generic or not, using unknown forces their hand.

Tobbe replied:
Closing this for now. Can revisit later if I/we want to

Tobbe reviewed (commented):
One comment specifically about `userAttributes`, and then a more generic comment about Auth method types.

Tobbe commented on __fixtures__/test-project/web/src/components/Contact/Contacts/Contacts.tsx:
```diff
@@ -27,19 +30,19 @@ const formatEnum = (values: string | string[] | null | undefined) => {
-  if (text && text.length > MAX_STRING_LENGTH) {
-    output = output.substring(0, MAX_STRING_LENGTH) + '...'
+const truncate = (value: string | number) => {
+  let output = value?.toString()
```
```suggestion
  const output = value?.toString()
```

dac09 replied:
Yup. correct, I'll fix it

jtoar commented on docs/docs/typescript/generated-types.md:
````diff
@@ -0,0 +1,157 @@
+
+For example, adding this `codegen.yml` to the root of your project will transform the names of the generated types to UPPERCASE:
+
+```yml title="codegen.yml"
````
This one isn't a huge deal, so let's just leave it as is at the moment and I'll file an issue on docusaurus's repo or whatever does the highlighting.

jtoar commented on docs/docs/typescript/generated-types.md:
````diff
@@ -0,0 +1,157 @@
+}
+```
+
+These types will also be handled automatically. But if you're returning a different Prisma model, you may need to write your own resolver type, as the type generator won't know how to map the Prisma type to the GraphQL return type.
````
```suggestion
These types will also be handled automatically. But if you're returning a different Prisma model (instead of something like the generic `OutOfStock` type we have here, which is just a message), you may need to write your own resolver type, as the type generator won't know how to map the Prisma type to the GraphQL return type.
```

jtoar commented on docs/docs/typescript/generated-types.md:
````diff
@@ -0,0 +1,157 @@
+}
+```
+
+These types help you by making sure you're returning an object in the shape of what you've defined in your SDL. Note that these types expect you to return the _complete_ type that you've defined in your SDL. You can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.
````
```suggestion
These types help you by making sure you're returning an object in the shape of what you've defined in your SDL. Note that these types expect you to return the _complete_ type that you've defined in your SDL. But you can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.
```

jtoar commented on docs/docs/typescript/generated-types.md:
```diff
@@ -0,0 +1,157 @@
+
+If you're curious, you can find the generated types in the `.redwood/types`, `web/types/graphql.d.ts`, and `api/types/graphql.d.ts` directories. Broadly speaking, Redwood generates the following types:
+
+1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, Cells, and layouts on the web side - and services, libs, etc. on the api side
```
```suggestion
1. ["mirror" types](https://www.typescriptlang.org/docs/handbook/module-resolution.html#virtual-directories-with-rootdirs) for your components, pages, layouts, etc. on the web side, and for your services, lib, etc. on the api side
```

jtoar reviewed (changes requested):
@dac09 couple more minor suggestions based on your answers to my questions from my last review. Then GTG on my end!

Tobbe commented on docs/docs/typescript/generated-types.md:
````diff
@@ -0,0 +1,157 @@
+}
+```
+
+These types help you by making sure you're returning an object in the shape of what you've defined in your SDL. Note that these types expect you to return the _complete_ type that you've defined in your SDL. You can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.
````
Was it here that we determined that the shape wasn't based on the SDL but rather on the Prisma schema?
```suggestion
These types help you by making sure you're returning an object in the shape of what you've defined in your Prisma schema. Note that these types expect you to return the _complete_ type that you've defined in your schema. You can just return the result of the Prisma query, and not have to worry about how, for example, a DateTime in Prisma maps to a String in GraphQL.
```

dac09 replied:
Both these statements are true. If the Model exists in Prisma, it'll use the Prisma model, if it doesn't it'll use the SDL model.

Tobbe replied:
Ok, can we make that clearer in the docs somehow?

Tobbe replied:
Looks great now!

Tobbe commented:
I created an issue for the prisma/sdl/resolvers thing I brought up for discussion on a CTM a while back (where I added `age` to the SDL to demonstrate a potential shortcoming in our types)

https://github.com/redwoodjs/redwood/issues/5982

dac09 commented:
I think I addressed all the comments here. @Tobbe, @jtoar 

But right now unable to regenerate the project (to remove the conflicts) - due to the problem introduced in https://github.com/redwoodjs/redwood/pull/5680 - where we can't regenerate the project (base64url module not found).

jtoar commented:
I think we ran into this before and the problem may have been that we don't copy deps over (line 233 below):

https://github.com/redwoodjs/redwood/blob/8884255ee92b8cf53f2322cb1107e9a25fa20e08/tasks/test-project/test-project#L219-L224

If you just pass `--rebuild-fixture`, we do three other link steps:
- `'[link] Building Redwood framework'`
- `'[link] Copying framework packages to project'`
- `'[link] Add rwfw project:copy postinstall'`

But we skip `'[link] Adding framework dependencies to project'`, so if a new dep gets added (in this case, `base64url`), things throw. Given that the CLI imports way too much at startup, I'm sure `base64url` is being required somehow, even though it's a web generator that's being run.

Not sure what the solution is right away yet; it may be as simple as enabling that step. I'll try a few things tomorrow.

cannikin commented:
> Not sure what the solution is right away yet; it may be as simple as enabling that step. I'll try a few things tomorrow.

Thanks @jtoar! Yeah I was freaking out earlier trying to figure out what's happening...if you look in redwood's `yarn.lock` the `base64url` dependency is in there, but when you look at the `yarn.lock` of the test project it's missing.

I'm working on a patch that'll move that dependency to the app itself, installed if/when you install WebAuthn along with dbAuth. 

There's one more issue that @dac09 is running into, where there's now a prompt when setting up dbAuth asking if you want to enable WebAuthn. This trips up the test-project runner as it doesn't know to answer the question. I was going to add a flag that you can include to skip the prompt completely.

cannikin commented:
HMM I thought I was in another PR and just resolved a couple of conflicts that weren't mine to resolve! I kept an `import type` that wasn't present in `main`, sorry about that!
